drivers/sensor/sensor_shell.c

```diff
@@ -264,45 +282,114 @@ static void sensor_shell_processing_callback(int result, uint8_t *buf, uint32_t
 		return;
 	}
 
-	rc = decoder->get_timestamp(buf, &timestamp);
-	if (rc != 0) {
-		shell_error(ctx->sh, "Failed to get fetch timestamp for '%s'", ctx->dev->name);
-		return;
-	}
-	shell_print(ctx->sh, "Got samples at %" PRIu64 " ns", timestamp);
+	for (int channel = 0; channel < SENSOR_CHAN_ALL; ++channel) {
+		uint32_t fit = 0;
+		size_t base_size;
+		size_t frame_size;
+		size_t channel_idx = 0;
+		uint16_t frame_count;
 
-	while (decoder->decode(buf, &fit, &cit, &channel, &q, 1) > 0) {
-		int8_t shift;
+		if (channel == SENSOR_CHAN_ACCEL_X || channel == SENSOR_CHAN_ACCEL_Y ||
+		    channel == SENSOR_CHAN_ACCEL_Z || channel == SENSOR_CHAN_GYRO_X ||
+		    channel == SENSOR_CHAN_GYRO_Y || channel == SENSOR_CHAN_GYRO_Z ||
+		    channel == SENSOR_CHAN_MAGN_X || channel == SENSOR_CHAN_MAGN_Y ||
+		    channel == SENSOR_CHAN_MAGN_Z || channel == SENSOR_CHAN_POS_DY ||
+		    channel == SENSOR_CHAN_POS_DZ) {
```

Comment:
There seems to be a lot of channel-specific additions like this throughout the PR. Have you measured the impact to code size?

Reply:
I have not, specifically because the roadmap is to eventually deprecate the old APIs and then all of these will go away. Happy to measure of you'd like.

Reply:
Can you show the rom report?

The 1kb rom increase is solely from this vertical decoder change is that right? Can you annotate this with the git sha's?

Reply:
[rom_report.base.txt](https://github.com/zephyrproject-rtos/zephyr/files/12621139/rom_report.base.txt)
[rom_report.diff.txt](https://github.com/zephyrproject-rtos/zephyr/files/12621143/rom_report.diff.txt)
[rom_report.new.txt](https://github.com/zephyrproject-rtos/zephyr/files/12621144/rom_report.new.txt)

Reply:
The diff is almost entirely in the sensors (since they need to support both code paths) and in the shell

Reply:
I expect some overhead in supporting old and new APIs simultaneously. My question was about the code impact in changing from horizontal to vertical decoding in the new API.

Reply:
@MaureenHelm @teburd OK... I did some more exploring and this increase is 100% a part of supporting 2 data representations. I took the icm42688_decoder.c as the example because we had a good before/after fully implemented decoder. The cost there comes from decoding to the old sensor value then to the q31. When (locally) I changed it directly to the q31 the decoder was actually 264B smaller (27.5%). It's not entirely a fair comparison because the old decoder would also benefit from exactly the same change (converting directly to q31 format).

I'm happy to make the change, but while we support 2 code paths on these sensors I think it's cleaner to use the common representation for readability. Later when we deprecate the old fetch/get we can optimize back down.

Reply:
I'm ok with this, I think mostly the code size should be about the same. Theoretically the only significant change would be the way the frame iterator becomes an index into the buffer to copy out. That really shouldn't be a big jump in code size.